src/lib.rs

```diff
@@ -129,18 +152,77 @@ impl StdError for Error {
         match *self {
             AlignmentMismatch { .. } => "Alignment Mismatch",
             LengthMismatch { .. } => "Length Mismatch",
+            CapacityMismatch { .. } => "Capacity Mismatch",
         }
     }
 }
 
-fn check_constraints<U>(data: &[u8]) -> Result<usize, Error>
+trait Empty {
+    fn is_empty(&self) -> bool;
+}
+
+impl<T> Empty for Vec<T> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl<T> Empty for [T] {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+trait Len {
+    fn len(&self) -> usize;
+}
+
+impl<T> Len for Vec<T> {
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+impl<T> Len for [T] {
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+trait Capacity {
+    fn capacity(&self) -> usize;
+}
+
+impl<T> Capacity for Vec<T> {
```

Comment:
This trait shouldn't be needed as you only ever need it for `Vec<T>`

Reply:
Okay :-).